src/main/java/teammates/logic/core/CoursesLogic.java

```diff
@@ -719,4 +719,27 @@ public boolean isCourseArchived(CourseAttributes course, InstructorAttributes in
         boolean isCourseArchived = (instructor.isArchived != null) ? instructor.isArchived : course.isArchived;
         return isCourseArchived;
     }
-}
+    
+    // TODO: Optimize extractActiveCourses() and extractArchivedCourses() to reduce the number of repeated calls of
+    // isCourseArchived(), which retrieves information from the database
+    
+    public List<CourseDetailsBundle> extractActiveCourses(List<CourseDetailsBundle> courseBundles, String googleId) {
+        List<CourseDetailsBundle> result = new ArrayList<CourseDetailsBundle>();
+        for (CourseDetailsBundle courseBundle : courseBundles) {
+            if (!isCourseArchived(courseBundle.course.id, googleId)) {
+                result.add(courseBundle);
+            }
+        }
+        return result;
+    }
+    
+    public List<CourseDetailsBundle> extractArchivedCourses(List<CourseDetailsBundle> courseBundles, String googleId) {
+        List<CourseDetailsBundle> result = new ArrayList<CourseDetailsBundle>();
+        for (CourseDetailsBundle courseBundle : courseBundles) {
+            if (isCourseArchived(courseBundle.course.id, googleId)) {
+                result.add(courseBundle);
+            }
+        }
+        return result;
+    }
```

Comment:
Sorry, just checking, but what was the reason for moving these here from *PageAction again?
I can't remember, and I can't think of any, since these methods can be placed anywhere at all.
Is this extracting required in multiple classes?